File: demoandroid/src/main/java/com/andretietz/retroauth/demo/LoginActivity.java
package com.andretietz.retroauth.demo;

import android.accounts.Account;
import android.os.Build.VERSION;
import android.os.Build.VERSION_CODES;
import android.os.Bundle;
import android.webkit.CookieManager;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.andretietz.retroauth.AuthenticationActivity;
import com.github.scribejava.apis.GitHubApi;
import com.github.scribejava.core.builder.ServiceBuilder;
import com.github.scribejava.core.model.OAuth2AccessToken;
import com.github.scribejava.core.model.OAuth2Authorization;
import com.github.scribejava.core.oauth.OAuth20Service;

import java.util.Locale;
import java.util.concurrent.Callable;

import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.moshi.MoshiConverterFactory;
import retrofit2.http.GET;
import retrofit2.http.Header;

public class LoginActivity extends AuthenticationActivity {

    private final OAuth20Service helper = new ServiceBuilder(ProviderGithub.CLIENT_ID)
            .apiSecret(ProviderGithub.CLIENT_SECRET)
            .callback(ProviderGithub.CLIENT_CALLBACK)
            .scope("user:email")
            .build(GitHubApi.instance());

    @Override
    protected void onCreate(Bundle icicle) {
        super.onCreate(icicle);
        setContentView(R.layout.activity_login);


        WebView webView = findViewById(R.id.webView);
        webView.loadUrl(helper.getAuthorizationUrl());
        webView.getSettings().setJavaScriptEnabled(true);
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                final OAuth2Authorization authorization = helper.extractAuthorization(url);
                String code = authorization.getCode();
                if (code == null) {
                    view.loadUrl(url);
                } else {
                    Single.fromCallable(new TokenVerifier(helper, code))

                            .subscribeOn(Schedulers.io())
                            .subscribe(result -> {
                                        Account account = createOrGetAccount(result.name);
                                        storeToken(
                                                account,
                                                getRequestedTokenType(),
                                                result.token.getAccessToken());
                                        finalizeAuthentication(account);
                                    },
                                    Throwable::printStackTrace);
                }
                return true;
            }
        });
    }

    @Override
    public void finish() {
        CookieManager cookieManager = CookieManager.getInstance();
        if (VERSION.SDK_INT < VERSION_CODES.LOLLIPOP) {
            //noinspection deprecation
            cookieManager.removeAllCookie();
        } else {
            cookieManager.removeAllCookies(null);
        }
        super.finish();
    }

    private static class TokenVerifier implements Callable<LoginResult> {

        private final OAuth20Service service;
        private final String code;



        private final GithubInfoService api = new Retrofit.Builder()
                .baseUrl("http://api.github.com/")
                .client(client())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(MoshiConverterFactory.create())
                .build().create(GithubInfoService.class);

        static OkHttpClient client() {
            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
            return new OkHttpClient.Builder()
                    .addInterceptor(interceptor)
                    .build();
        }

        TokenVerifier(OAuth20Service service, String code) {
            this.service = service;
            this.code = code;
        }

        @Override
        public LoginResult call() throws Exception {
            OAuth2AccessToken token = service.getAccessToken(code);
            UserInfo info = api.getUserInfo(String.format(Locale.US, "token %s", token.getAccessToken())).blockingGet();
            return new LoginResult(info.login, token);
        }
    }

    interface GithubInfoService {
        @GET("/user")
        Single<UserInfo> getUserInfo(@Header("Authorization") String token);
    }

    static class UserInfo {
        public String login;
    }

    private static class LoginResult {
        public final String name;
        public final OAuth2AccessToken token;

        LoginResult(String name, OAuth2AccessToken token) {
            this.name = name;
            this.token = token;
        }
    }
}

File: demoandroid/src/main/java/com/andretietz/retroauth/demo/LoginActivity.kt
package com.andretietz.retroauth.demo

import android.annotation.TargetApi
import android.os.Build
import android.os.Bundle
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import com.andretietz.retroauth.AndroidToken
import com.andretietz.retroauth.AuthenticationActivity
import com.github.scribejava.apis.FacebookApi
import com.github.scribejava.core.builder.ServiceBuilder
import com.github.scribejava.core.model.OAuth2AccessToken
import com.github.scribejava.core.oauth.OAuth20Service
import com.github.scribejava.httpclient.okhttp.OkHttpHttpClient
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_login.webView
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.http.GET
import retrofit2.http.Query
import timber.log.Timber
import java.util.concurrent.Callable
import java.util.concurrent.TimeUnit

class LoginActivity : AuthenticationActivity() {
  private val compositeDisposable = CompositeDisposable()
  private val helper = ServiceBuilder(FacebookAuthenticator.CLIENT_ID)
    .apiSecret(FacebookAuthenticator.CLIENT_SECRET)
    .callback(FacebookAuthenticator.CLIENT_CALLBACK)
    .withScope("email")
    .httpClient(OkHttpHttpClient())
    .build(FacebookApi.instance())

  private val authenticator by lazy { FacebookAuthenticator(application) }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_login)
    Timber.plant(Timber.DebugTree())

    webView.loadUrl(helper.authorizationUrl)
    @Suppress("UsePropertyAccessSyntax")
    webView.getSettings().setJavaScriptEnabled(true)
    webView.webViewClient = object : WebViewClient() {
      @Suppress("OverridingDeprecatedMember")
      override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
        val authorization = helper.extractAuthorization(url)
        val code = authorization.code
        if (code == null) {
          view.loadUrl(url)
        } else {
          val disposable = Single.fromCallable(TokenVerifier(helper, code))
            .subscribeOn(Schedulers.io())
            .subscribe({ result ->
              val account = createOrGetAccount(result.name)
              val expiryDate = TimeUnit.MILLISECONDS
                // expiry date - 30 seconds (network tolerance)
                .convert((result.token.expiresIn - 30).toLong(), TimeUnit.SECONDS)
                .plus(System.currentTimeMillis())
              storeToken(
                account,
                authenticator.tokenType,
                result.token.toAndroidToken()
              )
              finalizeAuthentication(account)
            }, { error -> Timber.e(error) })
          compositeDisposable.add(disposable)
        }
        return true
      }

      @Suppress("DEPRECATION")
      @TargetApi(Build.VERSION_CODES.N)
      override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
        return shouldOverrideUrlLoading(view, request.url.toString())
      }
    }
  }

  override fun onDestroy() {
    compositeDisposable.dispose()
    super.onDestroy()
  }

  private class TokenVerifier(private val service: OAuth20Service, private val code: String) : Callable<LoginResult> {
    private val api = Retrofit.Builder()
      .baseUrl("https://graph.facebook.com/")
      .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
      .addConverterFactory(MoshiConverterFactory.create())
      .build().create(FacebookInfoService::class.java)

    override fun call(): LoginResult {
      val token = service.getAccessToken(code)
      val info = api.getUserInfo("name,email", token.accessToken).blockingGet()
      return LoginResult(info.email, token)
    }
  }

  internal interface FacebookInfoService {
    @GET("v2.11/me")
    fun getUserInfo(@Query("fields") fields: String, @Query("access_token") token: String): Single<UserInfo>
  }

  internal class UserInfo {
    var name: String = ""
    var email: String = ""
  }

  private class LoginResult(val name: String, val token: OAuth2AccessToken)

  fun OAuth2AccessToken.toAndroidToken(): AndroidToken {
    return AndroidToken(this.accessToken, mapOf(
      FacebookAuthenticator.KEY_TOKEN_VALIDITY
        to TimeUnit.MILLISECONDS
        // expiry date - 30 seconds (network tolerance)
        .convert((this.expiresIn - 30).toLong(), TimeUnit.SECONDS)
        .plus(System.currentTimeMillis()).toString()
    ))

  }
}
